File: lintcode/95_validate-binary-search-tree/validate-binary-search-tree.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/validate-binary-search-tree
@Language: C++
@Datetime: 16-04-17 19:01
*/

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */
    bool isValidBST(TreeNode *root) {
        TreeNode *pre = NULL;
        return inorder_helper(root, pre);
    }
    
    bool inorder_helper(TreeNode *cur, TreeNode *&pre) {        
        if(!cur) return true;
        if(!inorder_helper(cur->left, pre)) return false;
        if(pre && pre -> val >= cur -> val) return false;
        pre = cur;
        return inorder_helper(cur->right, pre);
    }
};

File: lintcode/95_validate-binary-search-tree/validate-binary-search-tree.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/validate-binary-search-tree
@Language: Python
@Datetime: 16-04-17 19:11
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: True if the binary tree is BST, or false
    """
    def isValidBST(self, root):
        stk = []
        pre, cur = None, root
        while stk or cur:
            if cur:
                stk.append(cur)
                cur = cur.left
            else:
                cur = stk.pop()
                if pre and pre.val >= cur.val:
                    return False
                pre, cur = cur, cur.right

        return True